airflow/models/baseoperator.py

```diff
@@ -909,6 +909,7 @@ def __init__(
         task_display_name: str | None = None,
         logger_name: str | None = None,
         allow_nested_operators: bool = True,
+        call_on_kill_on_dagrun_timeout: bool = True,
```

Comment:
@RNHTTR can help clarify, but basically this change would allow tasks to run on_kill when a dagrun reaches timeout. The use case given in the issue is that "Some users would like for externally running workloads (e.g. snowflake, emr, bigquery, etc etc) to stop executing when a DAG run times out." . With this flag the user will have more control on what behavior they want when there's a timeout.

Reply:
This all stems from the fairly controversial idea of [setting running tasks to skipped if a DAG run reaches its timeout](https://github.com/apache/airflow/issues/30264).

There are a couple paths forward:

1. Change the behavior after a `dagrun_timeout` is reached to mark running TIs as failed.
1. Proceed with this PR
2. Encourage users to use an [on_skipped_callback](https://airflow.apache.org/docs/apache-airflow/stable/administration-and-deployment/logging-monitoring/callbacks.html#callback-types) for external systems that need to be shutdown when a task moves to the skipped state after a DAG run timeout.

Reply:
@ashb @RNHTTR  Checking in here. Do we want to move forward with this PR?

Reply:
@ashb Any thoughts on what's the best path forward?

Also, I mentioned in a comment here that there has been a significant change to the DAG properties. If we want to use this new `call_on_kill_on_dagrun_timeout`, where should it go now?